### PR TITLE
fraud: add peer that sent an invalid fraud proof to black list

### PR DESCRIPTION
**Pull request by vgonkivs**
Resolves https://github.com/celestiaorg/celestia-node/issues/954

### CODE REVIEW

File: fraud/service_test.go [hunk 1]
```diff
@@ -56,18 +60,216 @@ func TestService_Broadcast(t *testing.T) {
...
+	_, err = serviceA.Subscribe(BadEncoding)
+	require.NoError(t, err)
+
+	subsB, err := serviceB.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsB.Cancel()
+
+	subsC, err := serviceC.Subscribe(BadEncoding)
+	require.NoError(t, err)
+	defer subsC.Cancel()
```
Comment (Wondertan) [inline fraud/service_test.go:230]: Cancel for A?

Comment (Wondertan) [reply fraud/service_test.go:230]: Same for another test

File: fraud/service_test.go [hunk 2]
```diff
@@ -56,18 +60,216 @@ func TestService_Broadcast(t *testing.T) {
...
+	require.NoError(t, err)
+
+	_, err = subsB.Proof(ctx)
+	require.NoError(t, err)
+
+	newCtx, cancel := context.WithTimeout(ctx, time.Millisecond*500)
+	t.Cleanup(cancel)
+	_, err = subsC.Proof(newCtx)
+	require.Error(t, err)
+	require.True(t, bl.Contains(net.Hosts()[1].ID()))
```
Comment (Wondertan) [inline fraud/service_test.go:154]: Also, Add here a check that publisher A is not blacklisted

File: fraud/interface.go [hunk 3]
```diff
@@ -28,7 +30,7 @@ type Service interface {
 type Broadcaster interface {
 	// Broadcast takes a fraud `Proof` data structure that implements standard BinaryMarshal
 	// interface and broadcasts it to all subscribed peers.
-	Broadcast(context.Context, Proof) error
+	Broadcast(context.Context, Proof, ...pubsub.PubOpt) error
```
Comment (Wondertan) [inline fraud/interface.go:33]: Feel very conflicted about relying on PubSub structures in our interface. Is there a way to avoid it by publishing directly via pubsub accessed through service? We anyway convert the Service interface to a specific type, so that should be doable.

**Review by Wondertan — COMMENTED** [review]
Nice!

One last thing we should to care about is testing/mocking implementations of Proofs so that we can test network behaviours, like in this tests, without relying on a specific Proof type.

File: fraud/service_test.go (hunk 1 above)
Comment (vgonkivs) [reply fraud/service_test.go:230]: `subsA` is not created: `_, err = serviceA.Subscribe(BadEncoding)`

File: fraud/interface.go (hunk 3 above)
Comment (vgonkivs) [reply fraud/interface.go:33]: yes, sure.

File: fraud/service_test.go (hunk 1 above)
Comment (Wondertan) [reply fraud/service_test.go:230]: Well, it is created because you call the func, you just ignore the value

Comment (vgonkivs) [reply fraud/service_test.go:230]: technically, yes. but should we still cancel it in this case?

Comment (Wondertan) [reply fraud/service_test.go:230]: and technically we should cancel. We should be nice to our code ;)

**Comment by vgonkivs** [comment]
> One last thing we should to care about is testing/mocking implementations of Proofs so that we can test network behaviours, like in this tests, without relying on a specific Proof type.

Yes, I agree. But let's make it in scope of separate PR. IMO, it's not related to this one. Will create and post here an issue.

**Comment by vgonkivs** [comment]
https://github.com/celestiaorg/celestia-node/issues/971

File: fraud/service_test.go [hunk 4]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+	defer t.Cleanup(cancel)
+	// create mock network
+	net, err := mocknet.FullMeshLinked(3)
+	require.NoError(t, err)
+	bServ := mdutils.Bserv()
+
+	// create first fraud service that will broadcast incorrect Fraud Proof
+	serviceA, store1 := createServiceWithHost(ctx, t, net.Hosts()[0])
+
+	h, err := store1.GetByHeight(context.TODO(), 1)
```
Comment (Wondertan) [inline fraud/service_test.go:279]: 🚓

File: fraud/service_test.go [hunk 5]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
```
Comment (Wondertan) [inline fraud/service_test.go:179]: 🚓

**Review by Wondertan — APPROVED** [review]
Nice!

The further #971 should prettify things a bit more

File: fraud/service.go [hunk 6]
```diff
@@ -106,25 +107,25 @@ func (f *service) processIncoming(
...
-		"from", msg.ReceivedFrom.String(),
+		"from", from.String(),
 	)
 	msg.ValidatorData = proof
 	f.storesLk.RLock()
 	store, ok := f.stores[proofType]
+	if !ok {
+		store = initStore(proofType, f.ds)
+		f.stores[proofType] = store
+	}
```
Comment (renaynay) [inline fraud/service.go:124]: if you're going to do this, then you need Lock, not RLock/RUnlock

File: fraud/service_test.go (hunk 5 above)
Comment (renaynay) [reply fraud/service_test.go:179]: Lol +1

**Review by renaynay — CHANGES_REQUESTED** [review]
Everything looks generally okay but I don't understand the reasoning behind adding the `from peer.ID` param to the validator func instead of just getting the peer from the `msg.ReceivedFrom` field -- what does this change?

Also need to go through tests - they are quite intense so i'm gonna do another pass through just for the test

File: fraud/service_test.go [hunk 7]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+	f := func(ctx context.Context, from peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+		msg.ValidatorData = befp
+		return pubsub.ValidationAccept
+	}
+	fserviceB.pubsub.RegisterTopicValidator(getSubTopic(BadEncoding), f) //nolint:errcheck
+	bin, err := befp.MarshalBinary()
+	require.NoError(t, err)
+	topic, ok := fserviceA.topics[BadEncoding]
+	require.True(t, ok)
+	// we cannot avoid sleep because it helps to avoid flakiness
```
Comment (renaynay) [inline fraud/service_test.go:150]: What is this sleep for though? What are you waiting on?

File: fraud/service_test.go [hunk 8]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+	time.Sleep(time.Millisecond * 100)
+	err = topic.Publish(ctx, bin, pubsub.WithReadiness(pubsub.MinTopicSize(1)))
+	require.NoError(t, err)
+
+	_, err = subsB.Proof(ctx)
+	require.NoError(t, err)
+
+	newCtx, cancel := context.WithTimeout(ctx, time.Millisecond*500)
+	defer cancel()
+	_, err = subsC.Proof(newCtx)
```
Comment (renaynay) [inline fraud/service_test.go:160]:
```suggestion

	_, err = subsC.Proof(newCtx)
```

File: fraud/service_test.go [hunk 9]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+	fserviceA := serviceA.(*service)
+	require.NotNil(t, fserviceA)
+
+	// create second service that will receive and validate Fraud Proof
+	serviceB, _ := createServiceWithHost(ctx, t, net.Hosts()[1])
+
+	fserviceB := serviceB.(*service)
+	require.NotNil(t, fserviceB)
+
+	bl, err := pubsub.NewTimeCachedBlacklist(time.Hour * 1)
```
Comment (renaynay) [inline fraud/service_test.go:93]: In a longer test like this, especially where the use of the var is further below the definition, can we use full names like `blocklist` here?

File: fraud/service_test.go [hunk 10]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+	require.NoError(t, err)
+
+	newCtx, cancel := context.WithTimeout(ctx, time.Millisecond*500)
+	defer cancel()
+	_, err = subsC.Proof(newCtx)
+	require.Error(t, err)
+	require.False(t, bl.Contains(net.Hosts()[0].ID()))
+	require.True(t, bl.Contains(net.Hosts()[1].ID()))
+	// we cannot avoid sleep because it helps to avoid flakiness
+	time.Sleep(time.Millisecond * 100)
```
Comment (renaynay) [inline fraud/service_test.go:165]: Why sleep here too?

File: fraud/service_test.go [hunk 11]
```diff
@@ -56,18 +60,334 @@ func TestService_Broadcast(t *testing.T) {
...
+}
+
+func TestService_GossipingOfFaultBEFP(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer t.Cleanup(cancel)
```
Comment (renaynay) [inline fraud/service_test.go:170]: you don't need to `defer` for `t.Cleanup` - it's redundant

**Review by renaynay — COMMENTED** [review]
So the tests are quite long and hard to follow what's going on and why - it would be nice to clean them up a bit, reduce LOC as much as possible (for example, no need to define `bServ` if it's only used in one place, just pass `mdutils.Bserv()` directly as a param). The tests inside `sync_test.go` are really nice to follow, for example.

**Comment by vgonkivs** [comment]
msg.ReceivedFrom field is the address of node that have created a Fraud Proof. But it could be a case when other node receives a Fraud Proof and regossips it further

File: fraud/service_test.go (hunk 7 above)
Comment (vgonkivs) [reply fraud/service_test.go:150]: libp2p specific sleeps 😞 
Test will fail without it

File: fraud/service_test.go (hunk 10 above)
Comment (vgonkivs) [reply fraud/service_test.go:164]: to avoid flakiness

File: fraud/service_test.go [hunk 12]
```diff
@@ -352,7 +348,8 @@ func TestService_GossipingOfBEFP(t *testing.T) {
 	require.NoError(t, err)
 
 	newCtx, cancel := context.WithTimeout(ctx, time.Millisecond*100)
-	t.Cleanup(cancel)
+	defer cancel()
```
Comment (renaynay) [inline fraud/service_test.go:351]: Nit but let's be consistent with way we cancel context. I see some places use t.Cleanup and some defer

**Review by renaynay — APPROVED** [review]
